kommandir/bin/adept_openstack.py

```diff
@@ -0,0 +1,720 @@
+        self.os_rest = OpenstackREST()
+        self.name = name
+        self.vm_state = 'building'
+        self.power_state = 0
+
+        self.os_rest.compute_request('/flavors', 'flavors')
+        flavor = self.os_rest.child_search('name', FLAVOR)
+        logging.debug("Flavor %s is id %s", FLAVOR, flavor['id'])
+
+        # Faster for the server to search for this
+        image = self.os_rest.service_request('image',
+                                             '/v2/images?name=%s&status=active' % IMAGE,
+                                             'images')
+        # Validates only one image was found, not really searching
+        image = self.os_rest.child_search('name', IMAGE)
+        logging.debug("Image %s is id %s", IMAGE, image['id'])
+
+        user_data = ("#cloud-config\n"
+                     # Because I'm self-centered
+                     "timezone: US/Eastern\n"
+                     # We will configure our own filesystems/partitioning
+                     "growpart:\n"
+                     "    mode: off\n"
+                     # Don't add silly 'please login as' to .ssh/authorized_keys
+                     "disable_root: false\n"
+                     # Allow password auth in case it's needed
+                     "ssh_pwauth: True\n"
+                     # Import all ssh_authorized_keys (below) into these users
+                     "ssh_import_id: [root]\n"
+                     # public keys to import to users (above)
+                     "ssh_authorized_keys: %s\n"
+                     # Prevent creating the default, generic user
+                     "users:\n"
+                     "   - name: root\n"
+                     "     primary-group: root\n"
+                     "     homdir: /root\n"
+                     "     system: true\n" % auth_key_lines)
+        logging.debug("Userdata: %s", user_data)
+
+        server_json = dict(
+            name=name,
+            flavorRef=flavor['id'],
+            imageRef=image['id'],
+            user_data=b64encode(user_data)
+        )
+        logging.info("Submitting creation request")
+        self.os_rest.compute_request('/servers', 'server',
+                                     'post', post_json=dict(server=server_json))
+
+    def done(self):
+        """Return True if VM is active and powered up"""
+        if self.power_state == 1 and self.vm_state == 'active':
+            return True
+        else:
+            time.sleep(0.5)  # don't busy-wait
+            return False
+
+    def action(self):
+        """Collect details about VM status, return server's UUID"""
+        self.os_rest.server(name=self.name)
+        self.vm_state = self.os_rest.response_json['OS-EXT-STS:vm_state']
+        self.power_state = self.os_rest.response_json['OS-EXT-STS:power_state']
+        logging.info("     %s: %s, power %s",
+                     self.name, self.vm_state,
+                     self.POWERSTATES[self.power_state])
```

Comment:
It makes me uncomfortable that the POWERSTATES are hardcoded here, with no coordination with the server. If next week they decide to add state 8, this well-intentioned debug statement will cause a difficult-to-diagnose crash.